src/models/document/base-document-content.ts

```diff
@@ -1002,209 +1008,115 @@ export const BaseDocumentContentModel = types
           // and added to the document later
-          const newAttributes = sharedDataSet.dataSet.attributes?.map(a => {
-            return { ...a, id: updatedEntry.attributeIdMap[a.id] };
-          });
-          const newCases = sharedDataSet.dataSet.cases?.filter(c => c.__id__).map(c => (
-            c.__id__ && { ...c, __id__: updatedEntry.caseIdMap[c.__id__] }
-          ));
+          const sharedModel = getSharedDataSetSnapshotWithUpdatedIds(sharedDataSet, updatedIds);
           newSharedModelEntries.push({
             tiles: sharedModelEntry.tiles,
-            sharedModel: {
-              ...sharedDataSet,
-              id: updatedEntry.sharedModelId,
-              dataSet: {
-                ...sharedDataSet.dataSet,
-                id: updatedEntry.dataSetId,
-                attributes: newAttributes,
-                cases: newCases
-              }
-            }
+            sharedModel
           });
         }
       });
 
       // Update tile content with new shared model ids
       const tileIdMap: Record<string, string> = {};
-      const updatedTiles: any[] = [];
+      const updatedTiles: IDropTileItem[] = [];
       tiles.forEach(tile => {
         const oldContent = JSON.parse(tile.tileContent);
-        const tileContent = oldContent;
+        const tileContent = cloneDeep(oldContent);
         tileIdMap[tile.tileId] = uniqueId();
 
         // Find the shared models for this tile
-        const sharedDataSetEntries = sharedModelEntries.filter(entry => (
-          entry.tiles.map(t => t.id).includes(tile.tileId)
-        ));
+        const sharedDataSetEntries =
+          sharedModelEntries.filter(entry => entry.tiles?.map(t => t.id).includes(tile.tileId));
 
         // Update the tile's references to its shared models
-        if (tile.tileType === "Table") {
-          const columnWidths: Record<string, number> = {};
-          sharedDataSetEntries.forEach(sharedDataSetEntry => {
-            const originalSharedDataSetId = sharedDataSetEntry.sharedModel.id;
-            const attributeIdMap = updatedSharedModelMap[originalSharedDataSetId].attributeIdMap;
-            for (const entry of Object.entries(oldContent.content.columnWidths)) {
-              const originalId = entry[0];
-              const width = entry[1] as number;
-              if (width !== undefined && originalId && attributeIdMap[originalId]) {
-                  columnWidths[attributeIdMap[originalId]] = width;
-              }
-            }
-          });
-          tileContent.content = { ...oldContent.content, columnWidths };
+        if (tile.tileType === "Geometry") {
+          tileContent.content =
+            updateGeometryContentWithNewSharedModelIds(oldContent.content, sharedDataSetEntries, updatedSharedModelMap);
+        } else if (tile.tileType === "Table") {
+          tileContent.content =
+            updateTableContentWithNewSharedModelIds(oldContent.content, sharedDataSetEntries, updatedSharedModelMap);
```

Comment:
Doesn't necessarily need to be done as part of this PR, but one way to handle this generically would be to register an `updateContentWithNewSharedModelIds` function with the `TileContentInfo` and then any tile that needed to could register its corresponding function.